File: scripts/shared/trends_inserter.py
"""
Trends Inserter - VERSÃO CORRIGIDA com source_id normalizado
"""

import json
from typing import Any, Dict, List


class TrendsInserter:
    """Classe genérica para inserir trends com source_id normalizado"""

    def __init__(self, conn):
        self.conn = conn
        self._source_cache = {}  # Cache de source_name -> source_id
        self._load_sources()

    def _load_sources(self):
        """Carrega mapeamento de sources para cache"""
        with self.conn.cursor() as cur:
            cur.execute("SELECT id, name FROM sofia.sources WHERE active = true")
            self._source_cache = {row[1]: row[0] for row in cur.fetchall()}

    def _get_source_id(self, source_name: str) -> int:
        """Obtém source_id do cache ou cria novo"""
        if source_name in self._source_cache:
            return self._source_cache[source_name]

        # Source não existe, criar
        with self.conn.cursor() as cur:
            cur.execute(
                """
                INSERT INTO sofia.sources (name, category, description)
                VALUES (%s, 'tech_trends', %s)
                ON CONFLICT (name) DO UPDATE SET name = EXCLUDED.name
                RETURNING id
            """,
                (source_name, f"Auto-created source: {source_name}"),
            )
            source_id = cur.fetchone()[0]
            self.conn.commit()

            # Atualizar cache
            self._source_cache[source_name] = source_id
            return source_id

    def insert(self, source: str, name: str, **kwargs):  # Nome da fonte (será convertido para source_id)
        """
        Insert com source_id normalizado

        Args:
            source: Nome da fonte (github, stackoverflow, npm, etc)
            name: Nome do trend
            **kwargs: Outros campos opcionais
        """

        # Obter source_id
        source_id = self._get_source_id(source)

        # Campos base
        fields = ["source_id", "name"]
        values = [source_id, name]
        placeholders = ["%s", "%s"]

        # Campos opcionais
        optional_fields = {
            "category",
            "trend_type",
            "score",
            "rank",
            "stars",
            "forks",
            "views",
            "mentions",
            "growth_rate",
            "period_start",
            "period_end",
            "metadata",
        }

        for field, value in kwargs.items():
            if field in optional_fields and value is not None:
                fields.append(field)

                if field == "metadata" and isinstance(value, dict):
                    values.append(json.dumps(value))
                else:
                    values.append(value)

                placeholders.append("%s")

        # Monta query
        fields_str = ", ".join(fields)
        placeholders_str = ", ".join(placeholders)

        update_fields = [f for f in fields if f not in ("source_id", "name", "period_start")]
        update_str = ", ".join([f"{f} = EXCLUDED.{f}" for f in update_fields])

        query = f"""
        INSERT INTO sofia.tech_trends ({fields_str})
        VALUES ({placeholders_str})
        ON CONFLICT (source_id, name, period_start) DO UPDATE SET
            {update_str},
            collected_at = NOW()
        """

        with self.conn.cursor() as cur:
            cur.execute(query, values)
        self.conn.commit()

    def batch_insert(self, trends: List[Dict[str, Any]]):
        """Batch insert"""
        for trend in trends:
            source = trend.pop("source")
            name = trend.pop("name")
            self.insert(source, name, **trend)


# Uso:
"""
inserter = TrendsInserter(conn)

# Source é convertido automaticamente para source_id
inserter.insert(source='github', name='react', stars=50000)
inserter.insert(source='npm', name='vue', score=95)
"""

File: scripts/shared/trends-inserter.ts
/**
 * Generic Trends Inserter - TypeScript
 * Insere qualquer tipo de trend em tech_trends com um único método
 */

import { Pool, PoolClient } from 'pg';

export interface TrendData {
    source: 'github' | 'stackoverflow' | 'npm' | 'pypi' | string;
    name: string;
    category?: string;
    trend_type?: string;
    score?: number;
    rank?: number;
    stars?: number;
    forks?: number;
    views?: number;
    mentions?: number;
    growth_rate?: number;
    period_start?: Date;
    period_end?: Date;
    metadata?: Record<string, any>;
}

export class TrendsInserter {
    private pool: Pool;

    constructor(pool: Pool) {
        this.pool = pool;
    }

    /**
     * Insert genérico - aceita qualquer combinação de campos
     */
    async insert(trend: TrendData, client?: PoolClient): Promise<void> {
        const db = client || this.pool;

        // Campos base obrigatórios
        const fields: string[] = ['source', 'name'];
        const values: any[] = [trend.source, trend.name];
        const placeholders: string[] = ['$1', '$2'];
        let paramCount = 2;

        // Campos opcionais
        const optionalFields: (keyof TrendData)[] = [
            'category', 'trend_type', 'score', 'rank',
            'stars', 'forks', 'views', 'mentions', 'growth_rate',
            'period_start', 'period_end', 'metadata'
        ];

        // Adiciona campos que foram passados
        for (const field of optionalFields) {
            if (trend[field] !== undefined && trend[field] !== null) {
                fields.push(field);

                // Converte metadata para JSON
                if (field === 'metadata') {
                    values.push(JSON.stringify(trend[field]));
                } else {
                    values.push(trend[field]);
                }

                placeholders.push(`$${++paramCount}`);
            }
        }

        // Monta query dinamicamente
        const fieldsStr = fields.join(', ');
        const placeholdersStr = placeholders.join(', ');

        // Campos para UPDATE (todos exceto source, name, period_start)
        const updateFields = fields.filter(f => !['source', 'name', 'period_start'].includes(f));
        const updateStr = updateFields.map(f => `${f} = EXCLUDED.${f}`).join(', ');

        const query = `
      INSERT INTO sofia.tech_trends (${fieldsStr})
      VALUES (${placeholdersStr})
      ON CONFLICT (source, name, period_start) DO UPDATE SET
        ${updateStr},
        collected_at = NOW()
    `;

        await db.query(query, values);
    }

    /**
     * Batch insert de múltiplos trends
     */
    async batchInsert(trends: TrendData[]): Promise<void> {
        const client = await this.pool.connect();
        try {
            await client.query('BEGIN');
            for (const trend of trends) {
                await this.insert(trend, client);
            }
            await client.query('COMMIT');
        } catch (error) {
            await client.query('ROLLBACK');
            throw error;
        } finally {
            client.release();
        }
    }
}

// Exemplo de uso:
/*
const inserter = new TrendsInserter(pool);

// GitHub
await inserter.insert({
  source: 'github',
  name: 'langchain',
  category: 'AI',
  stars: 50000,
  forks: 5000,
  period_start: new Date(),
  metadata: { language: 'Python' }
});

// StackOverflow
await inserter.insert({
  source: 'stackoverflow',
  name: 'python',
  views: 1000000,
  mentions: 500,
  period_start: new Date()
});

// NPM
await inserter.insert({
  source: 'npm',
  name: 'react',
  score: 95.5,
  mentions: 10000,
  period_start: new Date()
});

// Batch
await inserter.batchInsert([
  { source: 'github', name: 'pytorch', stars: 60000 },
  { source: 'npm', name: 'vue', score: 90 },
]);
*/
